From 99944a6722d39690469b54697a01adebb1e30f4a Mon Sep 17 00:00:00 2001
From: KayYCHuang-wiwynn <kay_yc_huang@wiwynn.com>
Date: Tue, 6 May 2025 08:40:27 +0000
Subject: [PATCH] i3c: aspeed: Add api to reset the i3c device

---
 drivers/i3c/i3c_aspeed.c  | 9 +++++++++
 include/drivers/i3c/i3c.h | 1 +
 2 files changed, 10 insertions(+)

diff --git a/drivers/i3c/i3c_aspeed.c b/drivers/i3c/i3c_aspeed.c
index e8d6be174dd..536b519b5ce 100644
--- a/drivers/i3c/i3c_aspeed.c
+++ b/drivers/i3c/i3c_aspeed.c
@@ -2075,6 +2075,15 @@ static void i3c_rst_worker(struct k_work *work)
 		obj->rst_cb(obj->dev);
 }
 
+void i3c_aspeed_reset_and_callback(const struct device *dev)
+{
+	LOG_INF("[DEBUG KAY][Checkpoint 2] i3c_aspeed_reset_and_callback");
+    struct i3c_aspeed_obj *obj = DEV_DATA(dev);
+    i3c_aspeed_init(dev);
+    if (obj->rst_cb)
+        obj->rst_cb(dev);
+}
+
 int i3c_aspeed_master_send_entdaa(struct i3c_dev_desc *i3cdev)
 {
 	struct i3c_aspeed_obj *obj = DEV_DATA(i3cdev->bus);
diff --git a/include/drivers/i3c/i3c.h b/include/drivers/i3c/i3c.h
index a50ccf03f3d..14bb23f6704 100644
--- a/include/drivers/i3c/i3c.h
+++ b/include/drivers/i3c/i3c.h
@@ -271,3 +271,4 @@ int i3c_i2c_write(struct i3c_dev_desc *slave, uint8_t addr, uint8_t *buf, int le
 typedef void (*i3c_rst_cb_t)(const struct device *dev);
 void i3c_aspeed_hook_rst_cb(const struct device *dev, i3c_rst_cb_t cb);
 #define i3c_hook_rst_cb          i3c_aspeed_hook_rst_cb
+void i3c_aspeed_reset_and_callback(const struct device *dev);
-- 
2.25.1

